src/IO/Exporter/SpacetimeInterpolator.cpp

```diff
@@ -93,67 +93,109 @@ void SpacetimeInterpolator<Dim, Frame>::load_time_bounds(
-                       std::less_equal<double>{}) +
+  const auto upper_time =
+      std::upper_bound(lower_time, all_observation_values_.end(),
+                       time_bounds[1], std::less_equal<double>{}) +
       num_ghost_slices_ + 1;
-  const std::vector<double> obs_values(lower, upper);
-  const std::vector<size_t> obs_ids(
+  time_bounds_[0] = *(lower_time + num_ghost_slices_);
+  time_bounds_[1] = *(upper_time - num_ghost_slices_ - 1);
+  const auto lower_id =
       all_observation_ids_.begin() +
-          std::distance(all_observation_values_.begin(), lower),
+      std::distance(all_observation_values_.begin(), lower_time);
+  const auto upper_id =
       all_observation_ids_.begin() +
-          std::distance(all_observation_values_.begin(), upper));
-  const size_t num_obs = obs_ids.size();
+      std::distance(all_observation_values_.begin(), upper_time);
+  const auto num_obs = static_cast<size_t>(std::distance(lower_id, upper_id));
   ASSERT(num_obs >= time_interpolation_order_ + 1,
          "The number of time slices must be at least the time interpolation "
          "order plus one (NumTimeSlices >= "
              << time_interpolation_order_ + 1 << ").");
 
+  if (interpolators_.empty()) {
+    // No need to lock because we are loading interpolators for the first time
+    // so no other thread can be accessing the data.
+    for (auto it = lower_id; it != upper_id; ++it) {
+      interpolators_.emplace_back(volume_files_or_glob_, subfile_name_,
+                                  ObservationId{*it}, tensor_components_);
+    }
+    return;
+  }
+
   // Drop interpolators that are not needed anymore. Do this early to free up
   // memory.
+  auto first_interpolator_obs_id_it =
+      std::find(all_observation_ids_.begin(), all_observation_ids_.end(),
+                interpolators_.front().obs_id());
+  ASSERT(first_interpolator_obs_id_it != all_observation_ids_.end(),
+         "The first interpolator is not in the list of observation IDs.");
+  const auto last_interpolator_obs_id_it =
+      first_interpolator_obs_id_it +
+      static_cast<std::ptrdiff_t>(interpolators_.size());
+  const auto num_erase_front =
+      std::distance(first_interpolator_obs_id_it, lower_id);
+  const auto num_erase_back =
+      std::distance(upper_id, last_interpolator_obs_id_it);
+  if (num_erase_front == 0 and num_erase_back == 0) {
+    // No need to drop or load interpolators
+    return;
+  }
   {
     // Lock while we are modifying the vector
     const std::unique_lock lock(interpolators_mutex_);
-    for (auto it = interpolators_.begin(); it != interpolators_.end(); ++it) {
-      if (std::find(obs_ids.begin(), obs_ids.end(), it->obs_id()) ==
-          obs_ids.end()) {
-        it = interpolators_.erase(it);
-        --it;
-      }
+    if (num_erase_front > 0) {
+      interpolators_.erase(interpolators_.begin(),
+                           interpolators_.begin() + num_erase_front);
```

Comment:
This can go past `end` if the new and old valid ranges are disjoint and far enough apart.  I think an easy fix is to first check something like
```c++
if (num_erase_front >= interpolators_.size() or num_erase_back >= interpoaltors_.size()) {
  interpolators_.clear();
  // copy the "interpolators_.empty()" block from above
}
```